Fix for Pod status in Node Explorer View #272

shreyu86 opened the pull request:
@itowlson this fixes the issue with the pod status in node explorer view. Wanted to open an PR for feedback to see if we should move to `JSON` based parsing of status as compared to column based parsing. Also wanted to get some clarification on use of [`getPods`](https://github.com/Azure/vscode-kubernetes-tools/blob/master/src/explorer.ts#L258) vs [`kubectl.asLines`](https://github.com/Azure/vscode-kubernetes-tools/blob/master/src/explorer.ts#L181) to get pod information as to which is the preferred method. Happy to refactor `getPods` to reflect pod status.

itowlson commented:
Good questions.  The nice thing about `getPods` is it returns typed data, so we get some checking of code that uses it.  Though we would need to add the status to the returned `PodInfo` type, as it's not there at the moment; and we'd need to confirm that we can map the JSON `status.phase` to the `kubectl` RUNNING/COMPLETED statuses.

But `KubernetesResourceFolder` already uses `asLines`, and this is probably more efficient than pulling back the full JSON document.  So I'm pretty comfortable with changing `KubernetesNodeResource` to also use `asLines`.  However, if we do this, it would be good to centralise the pod line parsing logic - with this PR we semi-duplicate it (because of the different indexes?), and that could lead to maintenance woes in future...

cc @brendandburns who wrote the original code and may have input

shreyu86 commented:
@itowlson Thanks for the feedback, +1 on centralizing the pod line parsing logic. The reason that this PR duplicates it because of the extra `namespace` column being returned because of the `--all-namespaces` flag.

itowlson commented:
Yeah, I tried running the `kubectl` command (because I was surprised by the different indexes) and saw that the different indexes were needed.  I'm not sure what the best solution is but here's a possible idea:

* Define a `PodLineLayout` interface with members `NamespaceIndex`, `NameIndex`, etc.
* Define two instances of this interface - `NamespacedPodLineLayout` with the normal indexes, and `AllNamespacesPodLineLayout` with the +1 indexes.
* Define a `parsePodLine` method that takes a line and an instance of `PodLineLayout`.
* Call `parsePodLine` from the two different places with the different layout instances.

If you want to get fancy you could even infer the layouts from the header line... (which `asLines` currently discards but we could change that)

An _even fancier_ alternative would be to create a new method that uses the header line to convert each line into an object, so instead of returning `string[]` it would return `any[]` and the objects would have fields `NAMESPACE`, `NAME` etc. and you could index into them using string literals e.g. `pod['NAMESPACE]`, `pod['NAME']`.  That would break the dependence on positional parsing altogether...

Or there are other ways.  Just some ideas to consider.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
 
 function getIconForPodStatus(status: string): vscode.Uri {
-    if (status === "running" || status === "completed") {
+    if (status === "Running" || status === "Completed") {
```
I'd suggest using a case-insensitive check so we don't have problems with possible different casings when the function is used from new places.

shreyu86 replied:
Sounds good. Will change this.

shreyu86 commented:
@itowlson made following changes: 
- modified `asLines` to retain header
- converted each line into an object using the header column in `parsePodLine` breaking dependence on positional parsing altogether.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
     }
 }
 
+function parsePodLine(podInfo: string[]): any[] {
```
I think we can refine the return type a little, to `[key: string]: string` (which guarantees that it's a map from strings to strings, rather than involving arbitrary value data).

shreyu86 replied:
perfect, will do that.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
     }
 }
 
+function parsePodLine(podInfo: string[]): any[] {
```
This parses multiple lines, including the header, so should be called `parsePodLines` (plural).

shreyu86 replied:
sounds good.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
+    const headers = podInfo.shift();
+    const parsedHeaders = headers.toLowerCase().replace(/\s+/g, '|').split('|');
+    const pods = [];
+    podInfo.map((line) => {
```
This construct maps a `void` function over the string array, and uses the side effect of the `void` function pushing an object onto an object array.  Prefer `const pods = podInfo.map((line) => { ...; return podObject; });` - this says to map an object-returning function over the string array to produce an array of objects directly.

shreyu86 replied:
perfect, will change this.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
+    podInfo.map((line) => {
+        const podObject = {};
+        const bits = line.replace(/\s+/g, '|').split('|');
+        bits.map((columnValue, index) => {
```
Can this be done with `forEach`?  `map` implies the function is performing a transformation and you are interested in the return values of the transformation, whereas this is being done purely for its side-effects (to populate the `podObject`).

shreyu86 replied:
sounds good.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
+        const podObject = {};
+        const bits = line.replace(/\s+/g, '|').split('|');
+        bits.map((columnValue, index) => {
+            podObject[parsedHeaders[index]]= columnValue;
```
Space before `=`

shreyu86 replied:
fixing this.

itowlson commented on src/explorer.ts:
```diff
@@ -183,11 +198,13 @@ class KubernetesResourceFolder extends KubernetesFolder {
         }
+        if (this.kind.abbreviation === "pod") {
+            const parsedPods = parsePodLine(childrenLines.result);
+            return parsedPods.map((line) => {
```
The range variable is now a pod, not a line.

shreyu86 replied:
will make this change.

itowlson commented on src/explorer.ts:
```diff
@@ -183,11 +198,13 @@ class KubernetesResourceFolder extends KubernetesFolder {
+        if (this.kind.abbreviation === "pod") {
+            const parsedPods = parsePodLine(childrenLines.result);
+            return parsedPods.map((line) => {
+                return new KubernetesResource(this.kind, line.name, { status: line.status });
```
Interesting, I would have thought the header names were capitalised - does that not matter?

shreyu86 replied:
I changed the header check to be case sensitive, will change this as the check for status has to be case insensitive.

itowlson commented on src/kubectl.ts:
```diff
@@ -219,7 +219,7 @@ async function asLines(context: Context, command: string): Promise<Errorable<str
     if (shellResult.code === 0) {
         let lines = shellResult.stdout.split('\n');
-        lines.shift();
+        // lines.shift();
```
Won't this break other consumers of `asLines` who expect the headers to have been stripped off?

itowlson commented on src/kubectl.ts:
```diff
@@ -219,7 +219,6 @@ async function asLines(context: Context, command: string): Promise<Errorable<str
     const shellResult = await invokeAsync(context, command);
     if (shellResult.code === 0) {
         let lines = shellResult.stdout.split('\n');
-        lines.shift();
```
I think removing this line will break other consumers of `asLines`, which expect the header line to have been removed.

itowlson replied:
I suggest we should either:

* Have two functions, `asLines` and `asLinesWithHeader` (both returning string arrays)
* Have two functions, `asLines` (unparsed) and `fromLines` (parsed), the former returning a string array and the latter an object array
* Have a `keepHeaderLine` argument (requires changing all calls to specify whether they want to keep the header line)
* Return a `{ headerLine: string; resourceLines: string[] }` instead of a `string` (requires changing all calls to discard the header line, which I know could be done with an `unshift` but this strategy gives us compiler checking)

I quite like the second option as it minimises changes elsewhere in the codebase, and as the parsed `fromLines` is probably going to be handy going forward.  But not quite sure, and naming is going to be tricky...

shreyu86 replied:
Sounds good will go for second option.

itowlson commented on src/explorer.ts:
```diff
@@ -20,13 +20,28 @@ export function createKubernetesResource(kind: kuberesources.ResourceKind, id: s
     }
 }
 
+function parsePodLine(podInfo: string[]): any[] {
```
Looking at this function, I don't think it has anything to do with pods specifically - you have designed it to be completely generic.  So I think we should consider moving it into `kubectl` or `kubectlUtils` as a generic "get me a parsed representation of a tabular resource list" function.  In fact it would even work with Helm...

shreyu86 replied:
Cool, will move it to `kubectlUtils` if that is ok.

shreyu86 commented:
@itowlson will address your comments soon. I have made one last change to make sure that `KubernetesSelectorResource`s listing of pods can take advantage of the changes. I am relatively new to TypeScript and appreciate your feedback, thanks again.

shreyu86 commented:
@itowlson refactored code according to your feedback:

- Renamed `parsePodLine` as `parseLineOutput` and moved under `kubectlUtils` and replaced return `any[]` with `{ [key: string]: string }[]` 
- Added `fromLines` in `kubectl` to take advantage of `parseLineOutput` and swapped `asLines` (for pod status) in `explorer`

itowlson commented:
Thanks @shreyu86 for the quick fix!

shreyu86 commented on src/kubectl.ts:
```diff
@@ -227,6 +232,17 @@ async function asLines(context: Context, command: string): Promise<Errorable<str
+    const shellResult = await invokeAsync(context, command);
+    if (shellResult.code === 0) {
+        let lines = shellResult.stdout.split('\n');
+        lines = lines.filter((l) => l.length > 0);
```
@itowlson I missed checking for empty response here, which will break `parseLineOutput`.
Thinking about following solution or open to other solutions: 
```typescript
let parsedOutput = [];
        if (lines.length > 0) {
            parsedOutput = parseLineOutput(lines);
        }
        return { succeeded: true, result: parsedOutput};
```

itowlson replied:
Good catch.  You're right, if there are no resources then there is no header line.  I'd actually consider putting the fix into `kubectlUtils.parseLineOutput` -  you do `const headers = lineOutput.shift()` to get the header line, and checking `if (!headers)` should solve the problem nicely (and play better with the typechecker once Project Strict Nulls gets that far) - I am pretty sure `shift` returns `undefined` if there are no array elements...?

shreyu86 replied:
sounds good, testing this approach.
